File: backend/app/schemas/investment_policy.py
"""Investment policy schemas -- global and per-symbol policies."""

from __future__ import annotations

from typing import Literal

from pydantic import AliasChoices, BaseModel, ConfigDict, Field, field_validator, model_validator


RiskProfile = Literal["conservative", "balanced", "aggressive_growth"]
AddStyle = Literal["left_side_add", "pullback_add", "right_side_confirm", "batch_add"]
AssetRole = Literal[
    "core_growth",
    "faith_holding",
    "satellite_growth",
    "speculative",
    "btc_proxy",
    "cash_like",
    "index_etf",
    "watchlist",
    "forbidden",
    "unknown",
]
Conviction = Literal["high", "medium", "low"]
AiReviewStatus = Literal["unknown", "reasonable", "questionable", "risky"]


def normalize_policy_symbol(symbol: str) -> str:
    """Normalize symbol: strip, uppercase, remove exchange suffix."""
    value = str(symbol or "").strip().upper()
    if "." in value:
        value = value.split(".", 1)[0]
    return value.strip()


class GlobalInvestmentPolicyBase(BaseModel):
    """Base fields for global investment policy."""

    risk_profile: RiskProfile = "balanced"
    target_annual_return_pct: float | None = Field(default=None, ge=0, le=5)
    max_drawdown_tolerance_pct: float | None = Field(default=None, ge=0, le=1)
    allow_concentrated_position: bool = False
    allow_single_position_over_20_pct: bool = False
    allow_leverage: bool = False
    cash_reserve_pct: float | None = Field(default=None, ge=0, le=1)
    preferred_add_styles: list[AddStyle] = Field(default_factory=list)
    preferred_sell_style: str = ""
    holding_period: str = ""
    notes: str = ""
    enabled: bool = True


class GlobalInvestmentPolicyUpsert(GlobalInvestmentPolicyBase):
    """Request body for creating/updating global policy."""

    pass


class GlobalInvestmentPolicy(GlobalInvestmentPolicyBase):
    """Response model for global policy."""

    id: str = "global"
    policy_type: Literal["global"] = "global"
    created_at: str
    updated_at: str


class SymbolInvestmentPolicyBase(BaseModel):
    """Base fields for per-symbol investment policy."""

    model_config = ConfigDict(validate_by_name=True)

    symbol: str
    asset_role: AssetRole = "unknown"
    conviction: Conviction = "medium"
    user_preferred_target_position_pct: float | None = Field(
        default=None,
        ge=0,
        le=1,
        validation_alias=AliasChoices("user_preferred_target_position_pct", "target_position_pct"),
    )
    user_preferred_max_position_pct: float = Field(
        default=0.05,
        ge=0,
        le=1,
        validation_alias=AliasChoices("user_preferred_max_position_pct", "max_position_pct"),
    )
    user_preferred_min_position_pct: float = Field(
        default=0.0,
        ge=0,
        le=1,
        validation_alias=AliasChoices("user_preferred_min_position_pct", "min_position_pct"),
    )
    add_rules: list[str] = Field(default_factory=list)
    no_add_triggers: list[str] = Field(default_factory=list)
    sell_triggers: list[str] = Field(default_factory=list)
    hard_constraints: list[str] = Field(default_factory=list)
    soft_preferences: list[str] = Field(default_factory=list)
    notes: str = ""
    enabled: bool = True
    ai_review_status: AiReviewStatus = "unknown"
    ai_review_summary: str | None = None
    ai_review_updated_at: str | None = None

    @field_validator("symbol")
    @classmethod
    def _normalize_symbol(cls, value: str) -> str:
        normalized = normalize_policy_symbol(value)
        if not normalized:
            raise ValueError("symbol is required")
        return normalized

    @field_validator(
        "add_rules",
        "no_add_triggers",
        "sell_triggers",
        "hard_constraints",
        "soft_preferences",
        mode="before",
    )
    @classmethod
    def _coerce_string_list(cls, value: object) -> list[str]:
        if value is None:
            return []
        if isinstance(value, str):
            return [line.strip() for line in value.splitlines() if line.strip()]
        if isinstance(value, list):
            return [str(item).strip() for item in value if str(item).strip()]
        raise ValueError("must be a list of strings")

    @model_validator(mode="after")
    def _validate_position_order(self) -> "SymbolInvestmentPolicyBase":
        target = self.user_preferred_target_position_pct
        if target is None:
            target = self.user_preferred_min_position_pct
        if self.user_preferred_min_position_pct > target or target > self.user_preferred_max_position_pct:
            raise ValueError("user preferred position percentages must satisfy min <= target <= max")
        return self

    @property
    def target_position_pct(self) -> float | None:
        """Backward-compatible alias."""
        return self.user_preferred_target_position_pct

    @property
    def max_position_pct(self) -> float:
        """Backward-compatible alias."""
        return self.user_preferred_max_position_pct

    @property
    def min_position_pct(self) -> float:
        """Backward-compatible alias."""
        return self.user_preferred_min_position_pct


class SymbolInvestmentPolicyUpsert(SymbolInvestmentPolicyBase):
    """Request body for creating/updating a symbol policy."""

    pass


class SymbolInvestmentPolicy(SymbolInvestmentPolicyBase):
    """Response model for a symbol policy."""

    id: str
    policy_type: Literal["symbol"] = "symbol"
    created_at: str
    updated_at: str


class SymbolInvestmentPolicyListResponse(BaseModel):
    """Response for listing symbol policies."""

    items: list[SymbolInvestmentPolicy]


class InvestmentPolicySeedDefaultsResponse(BaseModel):
    """Response for seeding default policies."""

    created: list[SymbolInvestmentPolicy]
    skipped: list[SymbolInvestmentPolicy]
    message: str

File: frontend/src/types/investmentPolicy.ts
/** Investment policy types. */

export type RiskProfile = 'conservative' | 'balanced' | 'aggressive_growth'
export type AddStyle = 'left_side_add' | 'pullback_add' | 'right_side_confirm' | 'batch_add'
export type AssetRole =
  | 'core_growth'
  | 'faith_holding'
  | 'satellite_growth'
  | 'speculative'
  | 'btc_proxy'
  | 'cash_like'
  | 'index_etf'
  | 'watchlist'
  | 'forbidden'
  | 'unknown'
export type Conviction = 'high' | 'medium' | 'low'
export type AiReviewStatus = 'unknown' | 'reasonable' | 'questionable' | 'risky'

export interface GlobalInvestmentPolicy {
  id: string
  policy_type: 'global'
  risk_profile: RiskProfile
  target_annual_return_pct: number | null
  max_drawdown_tolerance_pct: number | null
  allow_concentrated_position: boolean
  allow_single_position_over_20_pct: boolean
  allow_leverage: boolean
  cash_reserve_pct: number | null
  preferred_add_styles: AddStyle[]
  preferred_sell_style: string
  holding_period: string
  notes: string
  enabled: boolean
  created_at: string
  updated_at: string
}

export interface SymbolInvestmentPolicy {
  id: string
  policy_type: 'symbol'
  symbol: string
  asset_role: AssetRole
  conviction: Conviction
  user_preferred_target_position_pct: number | null
  user_preferred_max_position_pct: number
  user_preferred_min_position_pct: number
  add_rules: string[]
  no_add_triggers: string[]
  sell_triggers: string[]
  hard_constraints: string[]
  soft_preferences: string[]
  notes: string
  enabled: boolean
  ai_review_status: AiReviewStatus
  ai_review_summary: string | null
  ai_review_updated_at: string | null
  created_at: string
  updated_at: string
}

export interface SymbolInvestmentPolicyListResponse {
  items: SymbolInvestmentPolicy[]
}
